### PR TITLE
HADOOP-18750. Remove javax/servlet shading in hadoop-client-api

### PR DESCRIPTION
**Pull request by kamalsharma2**
### Description of PR
This PR aims to resolve [HADOOP-18750](https://issues.apache.org/jira/browse/HADOOP-18750)

### How was this patch tested?
Tested locally.

### CODE REVIEW

**Comment by hadoop-yetus** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime |  Logfile | Comment |
|:----:|----------:|--------:|:--------:|:-------:|
| +0 :ok: |  reexec  |   0m 37s |  |  Docker mode activated.  |
|||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  |  No case conflicting files found.  |
| +0 :ok: |  codespell  |   0m  0s |  |  codespell was not available.  |
| +0 :ok: |  detsecrets  |   0m  0s |  |  detect-secrets was not available.  |
| +0 :ok: |  xmllint  |   0m  0s |  |  xmllint was not available.  |
| +1 :green_heart: |  @author  |   0m  0s |  |  The patch does not contain any @author tags.  |
| -1 :x: |  test4tests  |   0m  0s |  |  The patch doesn't appear to include any new or modified tests. Please justify why no new tests are needed for this patch. Also please list what manual steps were performed to verify this patch.  |
|||| _ trunk Compile Tests _ |
| +1 :green_heart: |  mvninstall  |  33m 41s |  |  trunk passed  |
| +1 :green_heart: |  compile  |   0m 26s |  |  trunk passed with JDK Ubuntu-11.0.19+7-post-Ubuntu-0ubuntu120.04.1  |
| +1 :green_heart: |  compile  |   0m 19s |  |  trunk passed with JDK Private Build-1.8.0_362-8u372-ga~us1-0ubuntu1~20.04-b09  |
| +1 :green_heart: |  mvnsite  |   0m 24s |  |  trunk passed  |
| +1 :green_heart: |  javadoc  |   0m 29s |  |  trunk passed with JDK Ubuntu-11.0.19+7-post-Ubuntu-0ubuntu120.04.1  |
| +1 :green_heart: |  javadoc  |   0m 24s |  |  trunk passed with JDK Private Build-1.8.0_362-8u372-ga~us1-0ubuntu1~20.04-b09  |
| +1 :green_heart: |  shadedclient  |  56m 20s |  |  branch has no errors when building and testing our client artifacts.  |
|||| _ Patch Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   2m 48s |  |  the patch passed  |
| +1 :green_heart: |  compile  |   0m 15s |  |  the patch passed with JDK Ubuntu-11.0.19+7-post-Ubuntu-0ubuntu120.04.1  |
| +1 :green_heart: |  javac  |   0m 15s |  |  the patch passed  |
| +1 :green_heart: |  compile  |   0m 14s |  |  the patch passed with JDK Private Build-1.8.0_362-8u372-ga~us1-0ubuntu1~20.04-b09  |
| +1 :green_heart: |  javac  |   0m 14s |  |  the patch passed  |
| +1 :green_heart: |  blanks  |   0m  1s |  |  The patch has no blanks issues.  |
| +1 :green_heart: |  mvnsite  |   0m 16s |  |  the patch passed  |
| +1 :green_heart: |  javadoc  |   0m 14s |  |  the patch passed with JDK Ubuntu-11.0.19+7-post-Ubuntu-0ubuntu120.04.1  |
| +1 :green_heart: |  javadoc  |   0m 14s |  |  the patch passed with JDK Private Build-1.8.0_362-8u372-ga~us1-0ubuntu1~20.04-b09  |
| -1 :x: |  shadedclient  |  23m 35s |  |  patch has errors when building and testing our client artifacts.  |
|||| _ Other Tests _ |
| +1 :green_heart: |  unit  |   0m 17s |  |  hadoop-client-api in the patch passed.  |
| +1 :green_heart: |  asflicense  |   0m 36s |  |  The patch does not generate ASF License warnings.  |
|  |   |  88m 41s |  |  |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.43 ServerAPI=1.43 base: https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-5695/1/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/5695 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient codespell detsecrets xmllint |
| uname | Linux f7692c0f1ea3 4.15.0-206-generic #217-Ubuntu SMP Fri Feb 3 19:10:13 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev-support/bin/hadoop.sh |
| git revision | trunk / f2567a8946defd72fe4b822ec7c1f1ca6c0f93d2 |
| Default Java | Private Build-1.8.0_362-8u372-ga~us1-0ubuntu1~20.04-b09 |
| Multi-JDK versions | /usr/lib/jvm/java-11-openjdk-amd64:Ubuntu-11.0.19+7-post-Ubuntu-0ubuntu120.04.1 /usr/lib/jvm/java-8-openjdk-amd64:Private Build-1.8.0_362-8u372-ga~us1-0ubuntu1~20.04-b09 |
|  Test Results | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-5695/1/testReport/ |
| Max. process+thread count | 559 (vs. ulimit of 5500) |
| modules | C: hadoop-client-modules/hadoop-client-api U: hadoop-client-modules/hadoop-client-api |
| Console output | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-5695/1/console |
| versions | git=2.25.1 maven=3.6.3 |
| Powered by | Apache Yetus 0.14.0 https://yetus.apache.org |


This message was automatically generated.

**Review by ayushtkn — CHANGES_REQUESTED** [review]
The change is there since 3.0-alpha release, was added as part of HADOOP-11804, why Spark fails with 3.3.4 only?

Isn't this an incompatible change for Hadoop?

**Comment by jojochuang** [comment]
The changes breaks Hadoop client shading because the classpath prefix is not allowed in the shaded client jar.

@sunchao any ideas why this is an issue with Hadoop 3.3.4...?

**Comment by sunchao** [comment]
@jojochuang @ayushtkn I think this has been an issue for a while (see https://issues.apache.org/jira/browse/SPARK-40964 for the same bug report) too, but just hasn't got a lot of attention - I just saw this issue today.

Long time back we also encountered a similar issue related to `javax.servlet` (see the discussions [here](https://issues.apache.org/jira/browse/SPARK-33212?page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel&focusedCommentId=17289618#comment-17289618)). In the end we found a workaround in Spark itself.

> +2021-02-24 08:36:54,391 ERROR org.apache.spark.SparkContext: Error initializing SparkContext.
java.lang.IllegalStateException: class org.apache.hadoop.yarn.server.webproxy.amfilter.AmIpFilter is not a javax.servlet.Filter+

In summary I think it'd be helpful to exclude `javax.servlet` from relocation.

**Comment by amanraj2520** [comment]
+1 I think we can remove javax.servlet from relocation.

**Comment by jojochuang** [comment]
> In summary I think it'd be helpful to exclude javax.servlet from relocation.

Ok. But still it breaks the build. I suspect we need to relax the allowed classpath prefix rule. @kamalsharma2 can you check if that's the case and update PR? It would be this script that need to update: https://github.com/apache/hadoop/blob/trunk/hadoop-client-modules/hadoop-client-check-invariants/src/test/resources/ensure-jars-have-correct-contents.sh
(use HADOOP-17891 as reference: https://github.com/apache/hadoop/commit/b8f7c7527a7c33c204315a6ea615b4d3fd237744)

**Comment by hadoop-yetus** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime |  Logfile | Comment |
|:----:|----------:|--------:|:--------:|:-------:|
| +0 :ok: |  reexec  |   0m 36s |  |  Docker mode activated.  |
|||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  |  No case conflicting files found.  |
| +0 :ok: |  codespell  |   0m  0s |  |  codespell was not available.  |
| +0 :ok: |  detsecrets  |   0m  0s |  |  detect-secrets was not available.  |
| +0 :ok: |  xmllint  |   0m  0s |  |  xmllint was not available.  |
| +0 :ok: |  shelldocs  |   0m  0s |  |  Shelldocs was not available.  |
| +1 :green_heart: |  @author  |   0m  0s |  |  The patch does not contain any @author tags.  |
| +1 :green_heart: |  test4tests  |   0m  0s |  |  The patch appears to include 1 new or modified test files.  |
|||| _ trunk Compile Tests _ |
| +0 :ok: |  mvndep  |  20m 56s |  |  Maven dependency ordering for branch  |
| +1 :green_heart: |  mvninstall  |  21m 56s |  |  trunk passed  |
| +1 :green_heart: |  compile  |   0m 24s |  |  trunk passed with JDK Ubuntu-11.0.19+7-post-Ubuntu-0ubuntu120.04.1  |
| +1 :green_heart: |  compile  |   0m 22s |  |  trunk passed with JDK Private Build-1.8.0_362-8u372-ga~us1-0ubuntu1~20.04-b09  |
| +1 :green_heart: |  mvnsite  |   0m 50s |  |  trunk passed  |
| +1 :green_heart: |  javadoc  |   0m 48s |  |  trunk passed with JDK Ubuntu-11.0.19+7-post-Ubuntu-0ubuntu120.04.1  |
| +1 :green_heart: |  javadoc  |   0m 43s |  |  trunk passed with JDK Private Build-1.8.0_362-8u372-ga~us1-0ubuntu1~20.04-b09  |
| +1 :green_heart: |  shadedclient  |  23m 57s |  |  branch has no errors when building and testing our client artifacts.  |
|||| _ Patch Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 29s |  |  Maven dependency ordering for patch  |
| +1 :green_heart: |  mvninstall  |   3m 43s |  |  the patch passed  |
| +1 :green_heart: |  compile  |   0m 16s |  |  the patch passed with JDK Ubuntu-11.0.19+7-post-Ubuntu-0ubuntu120.04.1  |
| +1 :green_heart: |  javac  |   0m 16s |  |  the patch passed  |
| +1 :green_heart: |  compile  |   0m 16s |  |  the patch passed with JDK Private Build-1.8.0_362-8u372-ga~us1-0ubuntu1~20.04-b09  |
| +1 :green_heart: |  javac  |   0m 16s |  |  the patch passed  |
| +1 :green_heart: |  blanks  |   0m  0s |  |  The patch has no blanks issues.  |
| +1 :green_heart: |  mvnsite  |   0m 32s |  |  the patch passed  |
| +1 :green_heart: |  shellcheck  |   0m  0s |  |  No new issues.  |
| +1 :green_heart: |  javadoc  |   0m 30s |  |  the patch passed with JDK Ubuntu-11.0.19+7-post-Ubuntu-0ubuntu120.04.1  |
| +1 :green_heart: |  javadoc  |   0m 28s |  |  the patch passed with JDK Private Build-1.8.0_362-8u372-ga~us1-0ubuntu1~20.04-b09  |
| -1 :x: |  shadedclient  |  22m 16s |  |  patch has errors when building and testing our client artifacts.  |
|||| _ Other Tests _ |
| +1 :green_heart: |  unit  |   0m 17s |  |  hadoop-client-api in the patch passed.  |
| +1 :green_heart: |  unit  |   0m 16s |  |  hadoop-client-check-invariants in the patch passed.  |
| +1 :green_heart: |  asflicense  |   0m 36s |  |  The patch does not generate ASF License warnings.  |
|  |   | 101m 19s |  |  |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.43 ServerAPI=1.43 base: https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-5695/2/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/5695 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient codespell detsecrets xmllint shellcheck shelldocs |
| uname | Linux 1c67920e37cf 4.15.0-206-generic #217-Ubuntu SMP Fri Feb 3 19:10:13 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev-support/bin/hadoop.sh |
| git revision | trunk / ec5e60d5426c2953dd48138af8879b46326f47db |
| Default Java | Private Build-1.8.0_362-8u372-ga~us1-0ubuntu1~20.04-b09 |
| Multi-JDK versions | /usr/lib/jvm/java-11-openjdk-amd64:Ubuntu-11.0.19+7-post-Ubuntu-0ubuntu120.04.1 /usr/lib/jvm/java-8-openjdk-amd64:Private Build-1.8.0_362-8u372-ga~us1-0ubuntu1~20.04-b09 |
|  Test Results | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-5695/2/testReport/ |
| Max. process+thread count | 560 (vs. ulimit of 5500) |
| modules | C: hadoop-client-modules/hadoop-client-api hadoop-client-modules/hadoop-client-check-invariants U: hadoop-client-modules |
| Console output | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-5695/2/console |
| versions | git=2.25.1 maven=3.6.3 shellcheck=0.7.0 |
| Powered by | Apache Yetus 0.14.0 https://yetus.apache.org |


This message was automatically generated.

**Comment by hadoop-yetus** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime |  Logfile | Comment |
|:----:|----------:|--------:|:--------:|:-------:|
| +0 :ok: |  reexec  |   0m 35s |  |  Docker mode activated.  |
|||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  |  No case conflicting files found.  |
| +0 :ok: |  codespell  |   0m  0s |  |  codespell was not available.  |
| +0 :ok: |  detsecrets  |   0m  0s |  |  detect-secrets was not available.  |
| +0 :ok: |  xmllint  |   0m  0s |  |  xmllint was not available.  |
| +0 :ok: |  shelldocs  |   0m  0s |  |  Shelldocs was not available.  |
| +1 :green_heart: |  @author  |   0m  0s |  |  The patch does not contain any @author tags.  |
| +1 :green_heart: |  test4tests  |   0m  0s |  |  The patch appears to include 1 new or modified test files.  |
|||| _ trunk Compile Tests _ |
| +0 :ok: |  mvndep  |  21m 25s |  |  Maven dependency ordering for branch  |
| +1 :green_heart: |  mvninstall  |  21m  1s |  |  trunk passed  |
| +1 :green_heart: |  compile  |   0m 26s |  |  trunk passed with JDK Ubuntu-11.0.19+7-post-Ubuntu-0ubuntu120.04.1  |
| +1 :green_heart: |  compile  |   0m 23s |  |  trunk passed with JDK Private Build-1.8.0_362-8u372-ga~us1-0ubuntu1~20.04-b09  |
| +1 :green_heart: |  mvnsite  |   0m 50s |  |  trunk passed  |
| +1 :green_heart: |  javadoc  |   0m 49s |  |  trunk passed with JDK Ubuntu-11.0.19+7-post-Ubuntu-0ubuntu120.04.1  |
| +1 :green_heart: |  javadoc  |   0m 42s |  |  trunk passed with JDK Private Build-1.8.0_362-8u372-ga~us1-0ubuntu1~20.04-b09  |
| +1 :green_heart: |  shadedclient  |  21m 19s |  |  branch has no errors when building and testing our client artifacts.  |
|||| _ Patch Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 29s |  |  Maven dependency ordering for patch  |
| +1 :green_heart: |  mvninstall  |   3m  1s |  |  the patch passed  |
| +1 :green_heart: |  compile  |   0m 16s |  |  the patch passed with JDK Ubuntu-11.0.19+7-post-Ubuntu-0ubuntu120.04.1  |
| +1 :green_heart: |  javac  |   0m 16s |  |  the patch passed  |
| +1 :green_heart: |  compile  |   0m 15s |  |  the patch passed with JDK Private Build-1.8.0_362-8u372-ga~us1-0ubuntu1~20.04-b09  |
| +1 :green_heart: |  javac  |   0m 15s |  |  the patch passed  |
| +1 :green_heart: |  blanks  |   0m  0s |  |  The patch has no blanks issues.  |
| +1 :green_heart: |  mvnsite  |   0m 34s |  |  the patch passed  |
| +1 :green_heart: |  shellcheck  |   0m  0s |  |  No new issues.  |
| +1 :green_heart: |  javadoc  |   0m 28s |  |  the patch passed with JDK Ubuntu-11.0.19+7-post-Ubuntu-0ubuntu120.04.1  |
| +1 :green_heart: |  javadoc  |   0m 30s |  |  the patch passed with JDK Private Build-1.8.0_362-8u372-ga~us1-0ubuntu1~20.04-b09  |
| -1 :x: |  shadedclient  |  20m 43s |  |  patch has errors when building and testing our client artifacts.  |
|||| _ Other Tests _ |
| +1 :green_heart: |  unit  |   0m 19s |  |  hadoop-client-api in the patch passed.  |
| +1 :green_heart: |  unit  |   0m 18s |  |  hadoop-client-check-invariants in the patch passed.  |
| +1 :green_heart: |  asflicense  |   0m 38s |  |  The patch does not generate ASF License warnings.  |
|  |   |  96m 10s |  |  |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.43 ServerAPI=1.43 base: https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-5695/3/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/5695 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient codespell detsecrets xmllint shellcheck shelldocs |
| uname | Linux 53cf23effe2f 4.15.0-206-generic #217-Ubuntu SMP Fri Feb 3 19:10:13 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev-support/bin/hadoop.sh |
| git revision | trunk / ef9defadf710531709f2a7a060d7373b3f792b5f |
| Default Java | Private Build-1.8.0_362-8u372-ga~us1-0ubuntu1~20.04-b09 |
| Multi-JDK versions | /usr/lib/jvm/java-11-openjdk-amd64:Ubuntu-11.0.19+7-post-Ubuntu-0ubuntu120.04.1 /usr/lib/jvm/java-8-openjdk-amd64:Private Build-1.8.0_362-8u372-ga~us1-0ubuntu1~20.04-b09 |
|  Test Results | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-5695/3/testReport/ |
| Max. process+thread count | 728 (vs. ulimit of 5500) |
| modules | C: hadoop-client-modules/hadoop-client-api hadoop-client-modules/hadoop-client-check-invariants U: hadoop-client-modules |
| Console output | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-5695/3/console |
| versions | git=2.25.1 maven=3.6.3 shellcheck=0.7.0 |
| Powered by | Apache Yetus 0.14.0 https://yetus.apache.org |


This message was automatically generated.

**Comment by jojochuang** [comment]
I checked out the branch and ran the exact same command in precommit test.

`mvn verify -fae --batch-mode -am -pl hadoop-client-modules/hadoop-client-check-invariants -pl hadoop-client-modules/hadoop-client-check-test-invariants -pl hadoop-client-modules/hadoop-client-integration-tests -Dtest=NoUnitTests -Dmaven.javadoc.skip=true -Dcheckstyle.skip=true -Dspotbugs.skip=true`

The error is it cannot find the classpath:


> [INFO] Running org.apache.hadoop.example.ITUseMiniCluster
> [ERROR] Tests run: 2, Failures: 0, Errors: 2, Skipped: 0, Time elapsed: 19.938 s <<< FAILURE! - in org.apache.hadoop.example.ITUseMiniCluster
> [ERROR] useWebHDFS(org.apache.hadoop.example.ITUseMiniCluster)  Time elapsed: 19.632 s  <<< ERROR!
> java.lang.NoClassDefFoundError: javax/servlet/Servlet
> 	at org.apache.hadoop.http.HttpServer2$Builder.build(HttpServer2.java:493)
> 	at org.apache.hadoop.hdfs.server.namenode.NameNodeHttpServer.start(NameNodeHttpServer.java:150)
> 	at org.apache.hadoop.hdfs.server.namenode.NameNode.startHttpServer(NameNode.java:1064)
> 	at org.apache.hadoop.hdfs.server.namenode.NameNode.initialize(NameNode.java:875)
> 	at org.apache.hadoop.hdfs.server.namenode.NameNode.<init>(NameNode.java:1130)
> 	at org.apache.hadoop.hdfs.server.namenode.NameNode.<init>(NameNode.java:1105)
> 	at org.apache.hadoop.hdfs.server.namenode.NameNode.createNameNode(NameNode.java:1879)
> 	at org.apache.hadoop.hdfs.MiniDFSCluster.createNameNode(MiniDFSCluster.java:1397)
> 	at org.apache.hadoop.hdfs.MiniDFSCluster.configureNameService(MiniDFSCluster.java:1166)
> 	at org.apache.hadoop.hdfs.MiniDFSCluster.createNameNodesAndSetConf(MiniDFSCluster.java:1039)
> 	at org.apache.hadoop.hdfs.MiniDFSCluster.initMiniDFSCluster(MiniDFSCluster.java:971)
> 	at org.apache.hadoop.hdfs.MiniDFSCluster.<init>(MiniDFSCluster.java:591)
> 	at org.apache.hadoop.hdfs.MiniDFSCluster$Builder.build(MiniDFSCluster.java:530)
> 	at org.apache.hadoop.example.ITUseMiniCluster.clusterUp(ITUseMiniCluster.java:77)
> 	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
> 	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
> 	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
> 	at java.lang.reflect.Method.invoke(Method.java:498)
> 	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
> 	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
> 	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
> 	at org.junit.internal.runners.statements.RunBefores.invokeMethod(RunBefores.java:33)
> 	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:24)
> 	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
> 	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
> 	at org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100)
> 	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366)
> 	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103)
> 	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:63)
> 	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
> 	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
> 	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
> 	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
> 	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
> 	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
> 	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
> 	at org.apache.maven.surefire.junit4.JUnit4Provider.execute(JUnit4Provider.java:365)
> 	at org.apache.maven.surefire.junit4.JUnit4Provider.executeWithRerun(JUnit4Provider.java:273)
> 	at org.apache.maven.surefire.junit4.JUnit4Provider.executeTestSet(JUnit4Provider.java:238)
> 	at org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:159)
> 	at org.apache.maven.surefire.booter.ForkedBooter.invokeProviderInSameClassLoader(ForkedBooter.java:384)
> 	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:345)
> 	at org.apache.maven.surefire.booter.ForkedBooter.execute(ForkedBooter.java:126)
> 	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:418)
> Caused by: java.lang.ClassNotFoundException: javax.servlet.Servlet
> 	at java.net.URLClassLoader.findClass(URLClassLoader.java:387)
> 	at java.lang.ClassLoader.loadClass(ClassLoader.java:418)
> 	at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:355)
> 	at java.lang.ClassLoader.loadClass(ClassLoader.java:351)
> 	... 44 more

**Comment by hadoop-yetus** [comment]
:confetti_ball: **+1 overall**






| Vote | Subsystem | Runtime |  Logfile | Comment |
|:----:|----------:|--------:|:--------:|:-------:|
| +0 :ok: |  reexec  |   0m 41s |  |  Docker mode activated.  |
|||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  |  No case conflicting files found.  |
| +0 :ok: |  codespell  |   0m  0s |  |  codespell was not available.  |
| +0 :ok: |  detsecrets  |   0m  0s |  |  detect-secrets was not available.  |
| +0 :ok: |  xmllint  |   0m  0s |  |  xmllint was not available.  |
| +0 :ok: |  shelldocs  |   0m  0s |  |  Shelldocs was not available.  |
| +1 :green_heart: |  @author  |   0m  0s |  |  The patch does not contain any @author tags.  |
| +1 :green_heart: |  test4tests  |   0m  0s |  |  The patch appears to include 1 new or modified test files.  |
|||| _ trunk Compile Tests _ |
| +0 :ok: |  mvndep  |  17m 33s |  |  Maven dependency ordering for branch  |
| +1 :green_heart: |  mvninstall  |  21m 43s |  |  trunk passed  |
| +1 :green_heart: |  compile  |   0m 25s |  |  trunk passed with JDK Ubuntu-11.0.19+7-post-Ubuntu-0ubuntu120.04.1  |
| +1 :green_heart: |  compile  |   0m 24s |  |  trunk passed with JDK Private Build-1.8.0_362-8u372-ga~us1-0ubuntu1~20.04-b09  |
| +1 :green_heart: |  mvnsite  |   1m 42s |  |  trunk passed  |
| +1 :green_heart: |  javadoc  |   1m 48s |  |  trunk passed with JDK Ubuntu-11.0.19+7-post-Ubuntu-0ubuntu120.04.1  |
| +1 :green_heart: |  javadoc  |   1m 36s |  |  trunk passed with JDK Private Build-1.8.0_362-8u372-ga~us1-0ubuntu1~20.04-b09  |
| +1 :green_heart: |  shadedclient  |  24m 50s |  |  branch has no errors when building and testing our client artifacts.  |
|||| _ Patch Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 31s |  |  Maven dependency ordering for patch  |
| +1 :green_heart: |  mvninstall  |  11m 54s |  |  the patch passed  |
| +1 :green_heart: |  compile  |   0m 16s |  |  the patch passed with JDK Ubuntu-11.0.19+7-post-Ubuntu-0ubuntu120.04.1  |
| +1 :green_heart: |  javac  |   0m 16s |  |  the patch passed  |
| +1 :green_heart: |  compile  |   0m 16s |  |  the patch passed with JDK Private Build-1.8.0_362-8u372-ga~us1-0ubuntu1~20.04-b09  |
| +1 :green_heart: |  javac  |   0m 16s |  |  the patch passed  |
| +1 :green_heart: |  blanks  |   0m  0s |  |  The patch has no blanks issues.  |
| +1 :green_heart: |  mvnsite  |   1m  8s |  |  the patch passed  |
| +1 :green_heart: |  shellcheck  |   0m  0s |  |  No new issues.  |
| +1 :green_heart: |  javadoc  |   1m  2s |  |  the patch passed with JDK Ubuntu-11.0.19+7-post-Ubuntu-0ubuntu120.04.1  |
| +1 :green_heart: |  javadoc  |   1m  4s |  |  the patch passed with JDK Private Build-1.8.0_362-8u372-ga~us1-0ubuntu1~20.04-b09  |
| +1 :green_heart: |  shadedclient  |  24m  8s |  |  patch has no errors when building and testing our client artifacts.  |
|||| _ Other Tests _ |
| +1 :green_heart: |  unit  |   0m 19s |  |  hadoop-client-api in the patch passed.  |
| +1 :green_heart: |  unit  |   0m 19s |  |  hadoop-client-runtime in the patch passed.  |
| +1 :green_heart: |  unit  |   0m 18s |  |  hadoop-client-check-invariants in the patch passed.  |
| +1 :green_heart: |  unit  |   0m 19s |  |  hadoop-client-minicluster in the patch passed.  |
| +1 :green_heart: |  asflicense  |   0m 38s |  |  The patch does not generate ASF License warnings.  |
|  |   | 111m 26s |  |  |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.43 ServerAPI=1.43 base: https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-5695/4/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/5695 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient codespell detsecrets xmllint shellcheck shelldocs |
| uname | Linux ab1325972b65 4.15.0-212-generic #223-Ubuntu SMP Tue May 23 13:09:22 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev-support/bin/hadoop.sh |
| git revision | trunk / df3b25828e6d0f0d11172a4bb0ccebcbf68f40c9 |
| Default Java | Private Build-1.8.0_362-8u372-ga~us1-0ubuntu1~20.04-b09 |
| Multi-JDK versions | /usr/lib/jvm/java-11-openjdk-amd64:Ubuntu-11.0.19+7-post-Ubuntu-0ubuntu120.04.1 /usr/lib/jvm/java-8-openjdk-amd64:Private Build-1.8.0_362-8u372-ga~us1-0ubuntu1~20.04-b09 |
|  Test Results | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-5695/4/testReport/ |
| Max. process+thread count | 733 (vs. ulimit of 5500) |
| modules | C: hadoop-client-modules/hadoop-client-api hadoop-client-modules/hadoop-client-runtime hadoop-client-modules/hadoop-client-check-invariants hadoop-client-modules/hadoop-client-minicluster U: hadoop-client-modules |
| Console output | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-5695/4/console |
| versions | git=2.25.1 maven=3.6.3 shellcheck=0.7.0 |
| Powered by | Apache Yetus 0.14.0 https://yetus.apache.org |


This message was automatically generated.

**Comment by kamalsharma2** [comment]
@jojochuang thanks for pointing out the issue. I've removed the shading of javax.servlet from other hadoop client modules too, can you please review it?

**Comment by ayushtkn** [comment]
I am not sure about it, In my personal experience: shading without relocation often leads to strange issues if there are different versions of the same dependency is the classpath and this jar should be widely used.

Does the solution pointed by @sunchao helps? or at worst if there is only few classes which creates issues, can we exclude just those classes from relocation?

In any case, I am not sure about this. I will pass to other more experienced folks to take a call on this.
